crates/codegen/parser/src/code_generator.rs

```diff
@@ -0,0 +1,325 @@
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    path::PathBuf,
+};
+
+use codegen_schema::types::grammar::Grammar;
+use codegen_utils::context::CodegenContext;
+use inflector::Inflector;
+use proc_macro2::{Ident, TokenStream};
+use quote::{format_ident, quote};
+use semver::Version;
+
+use super::naming;
+
+#[derive(Clone, Debug, Default)]
+pub struct VersionedFunctionBody {
+    pub comment: Vec<String>,
+    pub versions: BTreeMap<Version, TokenStream>,
+}
+
+impl VersionedFunctionBody {
+    fn insert(&mut self, comment: Vec<String>, version: &Version, body: TokenStream) {
+        self.comment = comment;
+        self.versions.insert(version.clone(), body);
+    }
+
+    fn to_function_body(&self, first_version: &Version) -> TokenStream {
+        self.versions
+            .iter()
+            .rev()
+            .map(|(version, body)| {
+                let version_flag = format_ident!(
+                    "version_is_equal_to_or_greater_than_{}",
+                    version.to_string().replace(".", "_")
+                );
+                if version == first_version {
+                    quote! { { #body } }
+                } else {
+                    quote! { if self.#version_flag { #body } }
+                }
+            })
+            .reduce(|a, b| quote! { #a else #b })
+            .unwrap()
+    }
+
+    fn to_comment_string(&self) -> String {
+        self.comment
+            .iter()
+            .map(|s| format!("// {}", s))
+            .collect::<Vec<_>>()
+            .join("\n")
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct CodeGenerator {
+    pub token_kinds: BTreeMap<String, Option<String>>,
```

Comment:
Not sure I can find where values of the `token_kinds` map are currently used. Are we using the terminal values somewhere?

Reply:
Maybe right, but not this PR.

Reply:
Not important if it is going to be used in the next PR. But otherwise, I suggest using a `BTreeSet` instead.